### ISSUE TITLE
Help with understanding grad_clip parameters - is it a feature or bug

### ISSUE THREAD

**Issue by marselap** [open-mmlab/mmdetection#8659]
**Describe the bug**
Gradient clipping seems to work in some experiments, and does not work in other runs with same parameters (unstable). 

All experiment parameters (configuration file, code version, dataset) kept the same, the clipping works in some runs, and in other runs the gradient can have large values (larger than the `max_norm` parameter). Runs are in the same environment, on the same worker.  We are not using `Deterministic` or `seed`.

1. What command or script did you run?

`tools/train.py` + customized config file

2. Did you make any modifications on the code or config? Did you understand what you have modified?

Config changes: 
-  Added the cyclic lr scheduler with linear warm up + cyclic momentum scheduler
- Custom anchors
- albumentations based training dataset augmentation pipeline
- 1 GPU, base_batch_size=64
- grad_clip=dict(max_norm=35, norm_type=2)
Code changes:
Custom ClearML logging hook on top of mmcv clearmllogger. Only has after_train_epoch callback implemented, but spikes occur at any iteration.

3. What dataset did you use?

Custom datasets in coco format

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment information and paste it here.
collect_env.py

```
sys.platform: linux
Python: 3.8.12 (default, Oct 12 2021, 13:49:34) [GCC 7.5.0]
CUDA available: True
GPU 0: Tesla T4
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.11.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.5.2 (Git Hash a9302535553c73243c632ad3c4c80beec3d19a1e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.11.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.12.0
OpenCV: 4.6.0
MMCV: 1.5.2
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.3
MMDetection: 2.25.1+d49d92a 
```

2. You may add addition that may be helpful for locating the problem, such as
   - How you installed PyTorch \[e.g., pip, conda, source\]
   - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

PATH 
`/opt/conda/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`

LD_LIBRARY_PATH
`/usr/local/nvidia/lib:/usr/local/nvidia/lib64`

Docker file
```
ARG PYTORCH="1.11.0"
ARG CUDA="11.3"
ARG CUDNN="8"

FROM pytorch/pytorch:${PYTORCH}-cuda${CUDA}-cudnn${CUDNN}-devel

ENV TORCH_CUDA_ARCH_LIST="6.0 6.1 7.0 8.0+PTX"
ENV TORCH_NVCC_FLAGS="-Xfatbin -compress-all"
ENV CMAKE_PREFIX_PATH="$(dirname $(which conda))/../"

RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys A4B469963BF863CC \
    && apt-get update && apt-get install -y ffmpeg libsm6 libxext6 git ninja-build libglib2.0-0 libsm6 libxrender-dev libxext6 libcurl4 mlocate nano \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* 

# Install MMCV
RUN pip install --no-cache-dir --upgrade pip wheel setuptools
RUN pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu113
RUN pip install --no-cache-dir mmcv-full -f https://download.openmmlab.com/mmcv/dist/cu113/torch1.11.0/index.html

# Install ClearML & Tensorboard
RUN pip install clearml future tensorboard

# Install FiftyOne
RUN pip install --index-url https://@pypi.fiftyone.ai -U fiftyone

# Install MMDetection and requirements
RUN conda clean --all
#RUN git clone https://github.com/open-mmlab/mmdetection.git /mmdetection
#WORKDIR /mmdetection
#ENV FORCE_CUDA="1"
#RUN pip install --no-cache-dir -r requirements/build.txt
#RUN pip install --no-cache-dir -r requirements/albu.txt
#RUN pip install --no-cache-dir -r requirements/tests.txt
#RUN pip install --no-cache-dir -r requirements/runtime.txt

# RUN pip install --no-cache-dir -e .
```

**Error traceback**

No error, the training runs regularly, and no error traceback. But we have gradients larger than max_norm, causing spikes in loss functions as well.

**Comment by Czm369** [comment]
In my experiment, grad_clip can avoid NaN in loss very well. We welcome you to share your in-depth research.
